README.md
# API-OpenWeather

This is a small example using https://openweathermap.org/ API.

# Requirements

- yarn or npm

# Start developing

## Install dependencies

``` shell
yarn # or npm i
```

## Build TypeScript

``` shell
yarn build # or npm run build
```


CHANGELOG.md
# Change Log

All notable changes to this publication will be documented in this file.

## 1.2.0 - 2023-25-01

Add type for API Response Body

## 1.1.0 - 2023-25-01

Separated CSS

## 1.0.0 - 2023-25-01

Add TypeScript to the project. Review of the already existent code.

More addictions:
- Devide project in public and src directories;
- CHANGELOG;
- LICENSE;
- package.json;
- .gitignore


src/scripts.ts
/**
 * OpenWeather API Example
 * Copyright (C) 2023  João Rodrigues, João Torres
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

type APIResponse = {
  coord: {
    lon: number
    lat: number
  }
  weather: [
    {
      id: number
      main: string
      description: string
      icon: string
    }
  ]
  base: string
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    humidity: number
    sea_level: number
    grnd_level: number
  }
  visibility: number
  wind: {
    speed: number
    deg: number
    gust: number
  }
  rain: {
    "1h": number
    "3h": number
  }
  snow: {
    "1h": number
    "3h": number
  }
  clouds: {
    all: number
  }
  dt: number
  sys: {
    type: number
    id: number
    message: unknown
    country: string
    sunrise: number
    sunset: number
  }
  timezone: number
  id: number
  name: string
  cod: number
};

const api = new URL("https://api.openweathermap.org/data/2.5/weather");

const proxy = new URL("https://cors-anywhere.herokuapp.com/");

const KEY = "your key";

const form = document.querySelector("form.search") as HTMLFormElement;
const weatherGrid = document.querySelector(".weather");

navigator.geolocation.getCurrentPosition(async ({ coords }) => {
    const { latitude, longitude } = coords;

    const map = new URL("https://www.bing.com/maps/embed");
    map.searchParams.set("h", "400");
    map.searchParams.set("w", "500");
    map.searchParams.set("cp", `${latitude}~${longitude}`);
    map.searchParams.set("lvl", "19");
    map.searchParams.set("typ", "d");
    map.searchParams.set("sty", "r");
    map.searchParams.set("src", "SHELL");
    map.searchParams.set("FORM", "MBEDV8");

    const geolocationApi = new URL("https://api.openweathermap.org/data/2.5/weather");
    geolocationApi.searchParams.set("lat", latitude.toString());
    geolocationApi.searchParams.set("lon", longitude.toString());
    geolocationApi.searchParams.set("appid", KEY);

    if (sessionStorage.getItem("localWeather") == null) {
        const apicoords = await fetch(new URL(geolocationApi.toString(), proxy));

        if (apicoords.ok) {
            const data = await apicoords.json();
            const city = data.name;
            const html = `
                <div>
                    <p>Latitude: ${latitude}</p>
                    <p>Longitude: ${longitude}</p>
                    <p>Cidade: ${city}</p>
                    <iframe width="500" height="400" frameborder="0" src="${map}" scrolling="no"></iframe>
                </div>
                `;

            if (weatherGrid != undefined) {
                weatherGrid.innerHTML = html;
            }

            sessionStorage.setItem('localWeather', JSON.stringify(data));
        }
    } else {
        const data = sessionStorage.getItem('localWeather');

        if (data == null) {
            return;
        }

        const weatherData = JSON.parse(data);

        const city = weatherData.name;

        const html = `
        <div>
            <p>Latitude: ${latitude}</p>
            <p>Longitude: ${longitude}</p>
            <p>Cidade: ${city}</p>
            <iframe width="500" height="400" frameborder="0" src="${map}" scrolling="no"></iframe>
        </div>
        `

        if (weatherGrid != undefined) {
            weatherGrid.innerHTML = html;
        }
    }
});

async function fetchWeather(city: string) {
    if (sessionStorage.getItem(`${city.toLowerCase()}`) === null) {
        const resource = new URL (api.toString(), proxy);
        resource.searchParams.set("q", city);
        resource.searchParams.set("appid", KEY);

        const res = await fetch(resource);
        const data = await res.json();
        sessionStorage.setItem(`${city.toLowerCase()}`, JSON.stringify(data));
        return data;
    } else {
        const weatherData = JSON.parse(sessionStorage.getItem(`${city.toLowerCase()}`) ?? "");
        return weatherData;
    }
}

async function handleSubmit(event: SubmitEvent) {
    event.preventDefault();

    const weather_form = event.currentTarget as HTMLFormElement;

    if (weather_form == null) {
        return;
    }

    //weather_form.submit.disabled = true;

    const weather = await fetchWeather(weather_form["query"].value);

    //weather_form.submit.disabled = false;

    //fetchWeather(weather_form.query.value);
    displayWeather(weather); //here i can use all of the data the api gives me, weather.name, weather.clouds, weather.coord etc...
}

function kToC(value: number) {
    const conversao = Math.floor(value - 273.15);
    return conversao;
}

function displayWeather(tempo: APIResponse) {
    if (weatherGrid == undefined) {
        return;
    }

    const html = `
        <div>
            <h2>${tempo.name}</h2>
            <p>Country: ${tempo.sys.country}</p>
            <p>Temperature: ${kToC(tempo.main.temp)}Cº</p>
            <p>Wind speed: ${tempo.wind.speed} Km/h</p>
            <p>Description: ${tempo.weather[0].description}</p>
        </div>
    `;

    weatherGrid.innerHTML += html;
}

form?.addEventListener('submit', handleSubmit);


scripts.js
/*
! IMPORTANT - CORS -  cross origin resource sharing -
* a domain name is an origin Ex: github.com is an origin
* due to CORS, by default and security issues you can´t access data from a website to another website, it will give you an error, you can if it is the same wesite, ex.: 
* you can fetch data from github.com to help.github.com because it is the same origin and the origin means the same domain in this case 'github.com';
^ if you´re using a local file... the url in browser would be something like... c://asd/asd/index.html or file:// ... you get it, so the first step to fetch data from another website to your own is to your file(website) in a server, it can be localhost or domain, myweb.com for example,
^ the second step, there´s is nothing you can do by your hands, because the website you´re trying to get data needs to allow your website to get the data, and this can be done by a line of code in the server of the website that you´re trying to get data from, the line is Access-Control-Allow-Origin: (your website);
* Ex.: imagine that you want to allow myfriend.com to access data from your website you have to insert this line Access-Control-Allow-Origin: myfriend.com

& All of this is called CORS Policy

* Sometimes the error is because of the browser , if you´re using the browser to access the API, without the Access-Control-Allow-Origin you won´t go anywhere, solutions to this is using server, node.js, PHP, ruby etc...
*---------------------OR-------------------
* you can also use a proxy CORS, BUT don´t ever do it if it has sensitive data like passwords, adresses, emails, because a proxy CORS is nothing more than a server used to get data from the website you want, then send it to you, but they get to see your data and what you fetched for, so dont use this with passwords emails etc...
*/
//* variavel api contem o link da api
const api = 'http://api.openweathermap.org/data/2.5/weather';
//* usar CORS proxy para aceder aos dados da api
const proxy = 'https://cors-anywhere.herokuapp.com/'
//* in the variable key, use your openweather key, you can obtain it here https://openweathermap.org/
const key = 'your key';
//* seleciona o elemento form
const form = document.querySelector('form.search');
//* seleciona o element div
const weatherGrid = document.querySelector('.weather');

function kToC(value){
    const conversao = Math.floor(value - 273.15);
    return conversao;
}

navigator.geolocation.getCurrentPosition((position) => {
    console.log(position.coords.latitude, position.coords.longitude);
    const lat = position.coords.latitude;
    const long = position.coords.longitude;
    const map = `https://www.bing.com/maps/embed?h=400&w=500&cp=${lat}~${long}&lvl=19&typ=d&sty=r&src=SHELL&FORM=MBEDV8`
    console.log('lat -> '+ lat + ' long -> '+ long);
    const geolocationApi = `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${long}&appid=${key}`
    
    if(sessionStorage.getItem('localWeather') === null){
        const apicoords = fetch(`${proxy}${geolocationApi}`);
        apicoords.then(response => {
            response.json().then(data => {
                console.log(data);
                const city = data.name;
                const html = `
                <div>
                    <h2>Cidade: ${city}</h2>
                    <p>Latitude: ${lat}</p>
                    <p>Longitude: ${long}</p>
                    <p>Temperature: ${kToC(data.main.temp)}Cº</p>
                    <iframe width="500" height="400" frameborder="0" src="${map}" scrolling="no"></iframe>
                </div>
                `
                weatherGrid.innerHTML = html;
                sessionStorage.setItem('localWeather', JSON.stringify(data));
            });
        });
    }else{
        const weatherData = JSON.parse(sessionStorage.getItem('localWeather'));
        console.log(weatherData);
        const city = weatherData.name;
        const html = `
        <div>
            <h2>Cidade: ${city}</h2>
            <p>Latitude: ${lat}</p>
            <p>Longitude: ${long}</p>
            <p>Temperature: ${kToC(weatherData.main.temp)}Cº</p>
            <iframe width="500" height="400" frameborder="0" src="${map}" scrolling="no"></iframe>
        </div>
        `
        weatherGrid.innerHTML = html;
    }
    
});

//* esta funcao vai usar a variavel 

async function fetchWeather(city){
    const currentTimestamp = Date.now();
    const storedWeather = localStorage.getItem(`${city.toLowerCase()}`);
    if(storedWeather === null){
        //* variavel res é o resultado da consulta (parametro query) feita na api, o conteudo retornado pela api sera guardado na variavel res
        const res = await fetch(`${proxy}${api}?q=${city}&appid=${key}`);
        const data = await res.json();
        localStorage.setItem(`${city.toLowerCase()}`, JSON.stringify({data:data, timestamp: currentTimestamp}));
        return data;
    } else {
        const storedWeatherData = JSON.parse(storedWeather); // data of the weather available in the localstorage
        const storedTimestamp = storedWeatherData.timestamp;
        if(currentTimestamp - storedTimestamp > 3600 * 1000){ // more than 1 hour has passed since the last update
            //* variavel res e o resultado da consulta (parametro query) feita na api, o conteudo retornado pela api sera guardado na variavel res
            const res = await fetch(`${proxy}${api}?q=${city}&appid=${key}`);
            const data = await res.json();
            localStorage.setItem(`${city.toLowerCase()}`, JSON.stringify({data:data, timestamp: currentTimestamp}));
            return data;
        }else{
            return storedWeatherData;
        }
    }
}

async function handleSubmit(event){
    event.preventDefault();
    const weather_form = event.currentTarget;
    //*turn the form off
    weather_form.submit.disabled = true;
    //*submit the search
    const weather = await fetchWeather(weather_form.query.value);
    console.log(weather);
    weather_form.submit.disabled = false;
    
    //fetchWeather(weather_form.query.value);
    displayWeather(weather); //this is the data the api gives me, it will be passed as an argument to the displayweather funciton
}

function displayWeather(tempo){ //this param is the object returned by the api response, so i can select all of its properties, tempo.name, tempo.sys.country etc...
    console.log('creating HTML');
    const html = `
        <div>
            <h2>${tempo.name}</h2>
            <p>Country: ${tempo.sys.country}</p>
            <p>Temperature: ${kToC(tempo.main.temp)}Cº</p>
            <p>Wind speed: ${tempo.wind.speed} Km/h</p>
            <p>Description: ${tempo.weather[0].description}</p>
        </div>
    `;
    weatherGrid.innerHTML += html;
}

form.addEventListener('submit', handleSubmit);